geth/txqueue/txqueue.go

```diff
@@ -183,66 +173,51 @@ func (q *TxQueue) Enqueue(tx *common.QueuedTx) error {
 
 // Get returns transaction by transaction identifier
 func (q *TxQueue) Get(id common.QueuedTxID) (*common.QueuedTx, error) {
-	q.mu.RLock()
-	defer q.mu.RUnlock()
-
-	if tx, ok := q.transactions[id]; ok {
-		return tx, nil
+	tx, ok := q.transactions.get(id)
+	if !ok {
+		return nil, ErrQueuedTxIDNotFound
 	}
 
-	return nil, ErrQueuedTxIDNotFound
+	return tx, nil
+}
+
+// Set transaction by transaction identifier
+func (q *TxQueue) Set(id common.QueuedTxID, tx *common.QueuedTx) {
+	q.transactions.add(id, tx)
 }
 
 // Remove removes transaction by transaction identifier
 func (q *TxQueue) Remove(id common.QueuedTxID) {
-	q.mu.Lock()
-	defer q.mu.Unlock()
-
-	delete(q.transactions, id)
+	q.transactions.delete(id)
 }
 
 // StartProcessing marks a transaction as in progress. It's thread-safe and
 // prevents from processing the same transaction multiple times.
 func (q *TxQueue) StartProcessing(tx *common.QueuedTx) error {
-	q.mu.Lock()
-	defer q.mu.Unlock()
-
-	if tx.Hash != (gethcommon.Hash{}) || tx.Err != nil {
+	if tx.Hash() != (gethcommon.Hash{}) || tx.Error() != nil {
```

Comment:
This is also a race condition. Goroutine A can evaluate this condition as `false` and just after that, goroutine B will set `tx.hash` or `tx.error` and `tx.inProgress` to `false`. Now, goroutine A calls `tx.Start` and it will be successful, although, the transaction has been already processed.

You can fix that by moving `tx.Start()` before this condition. Then, you should be able to remove `if tx.Hash() != (gethcommon.Hash{}) || tx.Error() != nil` completely, I guess. Or you can move it to `QueuedTx.Start` and... you will end up with `QueuedTx.Start()` being identical to `TxQueue.StartProcessing()`.

Reply:
"you will end up with QueuedTx.Start() being identical to TxQueue.StartProcessing()" that is totaly perfect for has-a relations. I think you suggestion great!